### PR TITLE
Fix mypy typing for `utilities.debugging`

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?
Part of #7037. This PR disentangles large #7218. It fixes mypy typing in `utilities.device_parser.py`.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/debugging.py [hunk 1]
```diff
@@ -16,39 +16,45 @@
...
-        self.saved_test_losses = []
-        self.early_stopping_history = []
-        self.checkpoint_callback_history = []
-        self.events = []
-        self.saved_lr_scheduler_updates = []
-        self.train_dataloader_calls = []
-        self.val_dataloader_calls = []
-        self.test_dataloader_calls = []
-        self.dataloader_sequence_calls = []
+        self.saved_train_losses: List[Dict[str, Union[float, torch.Tensor, object]]] = []
```
Comment (awaelchli) [inline pytorch_lightning/utilities/debugging.py:47]: what's the object type for? would it not be typed as Any?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/debugging.py (hunk 1 above)
Comment (stancld) [reply pytorch_lightning/utilities/debugging.py]: Yes, it seems `Any` typing is appropriate for this.

**Comment by carmocca** [comment]
There's one mypy error

`pytorch_lightning/utilities/debugging.py:173: error: Item "None" of "Optional[ModelCheckpoint]" has no attribute "monitor"  [union-attr]`

**Comment by stancld** [comment]
> There's one mypy error
> 
> `pytorch_lightning/utilities/debugging.py:173: error: Item "None" of "Optional[ModelCheckpoint]" has no attribute "monitor" [union-attr]`

The return type of `trainer.checkpoint_callback` is defined as `Optional[ModelCheckpoint]` from
```
/pytorch_lightning/trainer/properties.py:383:    def checkpoint_callback(self) -> Optional[ModelCheckpoint]:
```

Can we here always assume `checkpoint_callback` is not `None`?

**Comment by carmocca** [comment]
> Can we here always assume checkpoint_callback is not None?

No, users can disable it by setting `Trainer(checkpoint_callback=False)`

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by Borda — APPROVED** [review]
